### ISSUE TITLE
[REVIEW]: dolfin-adjoint 2018.1: automated adjoints for FEniCS and Firedrake

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1292]
**Submitting author:** @sebastkm (<a href="http://orcid.org/0000-0002-8888-6568">Sebastian Mitusch</a>)
**Repository:** <a href="https://bitbucket.org/dolfin-adjoint/pyadjoint" target ="_blank">https://bitbucket.org/dolfin-adjoint/pyadjoint</a>
**Version:** 2018.1.1
**Editor:** @Kevin-Mattheus-Moerman
**Reviewers:** @chrisrichardson, @IgorBaratta
**Archive:** <a href="https://doi.org/10.5281/zenodo.3247690" target="_blank">10.5281/zenodo.3247690</a>

## Status

[![status](http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d/status.svg)](http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d"><img src="http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d/status.svg)](http://joss.theoj.org/papers/76b42b8582fce57ffc3e3a0b9ba5699d)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://bitbucket.org/dolfin-adjoint/pyadjoint" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@chrisrichardson, @IgorBaratta, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://bitbucket.org/dolfin-adjoint/pyadjoint">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 2018.1.1
- [x] **Authorship:** Has the submitting author (@sebastkm) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
---
### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://bitbucket.org/dolfin-adjoint/pyadjoint">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** 2018.1.1
- [x] **Authorship:** Has the submitting author (@sebastkm) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @chrisrichardson it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01292/joss.01292/10.21105.joss.01292.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon add @IgorBaratta as reviewer

**Comment by whedon** [comment]
OK, @IgorBaratta is now a reviewer

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm, @chrisrichardson, @IgorBaratta, this is where the review process takes place. Let me know if you have any questions (Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations, which will enable you to thick the boxes at the top of this review issue).

**Comment by IgorBaratta** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- http://doi.org/10.1137/120873558 is OK

MISSING DOIs

- https://doi.org/10.1145/2998441 may be missing for title: Firedrake: automating the finite element method by composing abstractions
- https://doi.org/10.1145/2566630 may be missing for title: Unified form language: A domain-specific language for weak formulations of partial differential equations

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm can you check the above DOI's? Thank you

**Comment by Kevin-Mattheus-Moerman** [comment]
@chrisrichardson, @IgorBaratta thanks for your review efforts. I can see several unticked boxes above. Are you opening issues on the project repository for these (if so please link to them here as well so we can easily keep track of progress)? If not could you summarize your concerns/issues here so @sebastkm knows what to work on? Thanks!

**Comment by IgorBaratta** [comment]
@Kevin-Mattheus-Moerman , I'm almost finishing with my review. 
I've opened [1 issue](https://bitbucket.org/dolfin-adjoint/pyadjoint/issues/106/joss-review-error-collecting-tests) and created [1 PR](https://bitbucket.org/dolfin-adjoint/pyadjoint/pull-requests/77/joss-review-add-missing-dois/diff) for  a small fix.

So far everything is looks pretty good. I hope to go through all the boxes above by friday.

**Comment by IgorBaratta** [comment]
@Kevin-Mattheus-Moerman 
All boxes are now ticked, and from my side, this is good to go. 
Just need to update the pdf to reflect the changes in the doi.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #1292 with the following error: 

   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    16    0    16    0     0     30      0 --:--:-- --:--:-- --:--:--    30
Error reading bibliography ./references.bib (line 30, column 1):
unexpected "d"
expecting space, ",", white space or "}"
Error running filter pandoc-citeproc:
Filter returned error status 1
Looks like we failed to compile the PDF

**Comment by sebastkm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01292/joss.01292/10.21105.joss.01292.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1137/120873558 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1145/2998441 is OK
- 10.1145/2566630 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@chrisrichardson thanks again for agreeing to review this work. Can you summarize where we stand at this point? What should @sebastkm be working on? If you have opened issues on the project repository can you link to them here? Thanks!

**Comment by labarba** [comment]
Looks like the first reviewer gave a [recommendation](https://github.com/openjournals/joss-reviews/issues/1292#issuecomment-479050427) to accept.

👋  @chrisrichardson — We're waiting for you to complete the review. Can you give us a status update? When do you think you might finish? Thanks!

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @chrisrichardson for ticking those final boxes.

**Comment by Kevin-Mattheus-Moerman** [comment]
@wedon generate pdf

**Comment by sebastkm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01292/joss.01292/10.21105.joss.01292.pdf)

**Comment by labarba** [comment]
What's the status here, @Kevin-Mattheus-Moerman?

**Comment by Kevin-Mattheus-Moerman** [comment]
@labarba thanks for the reminder. I think this looks good and I'll proceed with the final steps.

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm at this point can you please do the following: 

- [ ] Minor suggestion, in your paper please consider using `[www.dolfin-adjoint.org](www.dolfin-adjoint.org)` such that it renders as a hyperlink in your paper. 
- [ ] Can you archive a copy of the reviewed software in [ZENODO](https://zenodo.org/)? Can you list the DOI of the archived version here when you have it? 
- [ ] Can you provide the version tag for the reviewed/archived package?

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm :point_up: any luck processing these last steps? Let me know if you have questions.

**Comment by sebastkm** [comment]
Done!
The DOI is: 10.5281/zenodo.3247690
The version tag is 2018.1.1

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.3247690 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3247690" target="_blank">10.5281/zenodo.3247690</a> is the archive.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set version as 2018.1.1

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 2018.1.1 as version

**Comment by whedon** [comment]
OK. 2018.1.1 is the version.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01292/joss.01292/10.21105.joss.01292.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm some final points: 

- [ ] Can you update the ZENODO meta data such that the archive title matches the JOSS paper title? 
- [ ] Did you see my minor suggestion to use this (rather than just `www.dolfin-adjoint.org`) in your paper such that the website renders as a hyperlink: `[www.dolfin-adjoint.org](www.dolfin-adjoint.org)`.

**Comment by sebastkm** [comment]
The updated joss paper is on a branch (or tag) `2018.1.1`. The one on master-branch is outdated at the moment. I could also update the master branch version, if it is not possible to use the 2018.1.1 version.

I've updated the ZENODO meta data now.

**Comment by Kevin-Mattheus-Moerman** [comment]
@sebastkm yes updating the master branch version would be easiest if that is okay. I think that is where `@whedon` builds the paper from.

**Comment by sebastkm** [comment]
Done

**Comment by sebastkm** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01292/joss.01292/10.21105.joss.01292.pdf)

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @sebastkm :rocket:

**Comment by Kevin-Mattheus-Moerman** [comment]
@openjournals/joss-eics This submission is ready to be accepted

**Comment by danielskatz** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- 10.1137/120873558 is OK
- 10.1007/978-3-642-23099-8 is OK
- 10.1145/2998441 is OK
- 10.1145/2566630 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/777

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/777, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
👋 @sebastkm - I've suggested some language changes in the paper in https://bitbucket.org/dolfin-adjoint/pyadjoint/pull-requests/89/papermd-edited-online-with-bitbucket/diff

**Comment by sebastkm** [comment]
Thanks! I have now merged the changes.

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by danielskatz** [comment]
Thanks to @Kevin-Mattheus-Moerman for editing and @chrisrichardson and @IgorBaratta for reviewing!

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1140948715575939072) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/779
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01292](https://doi.org/10.21105/joss.01292)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01292/status.svg)](https://doi.org/10.21105/joss.01292)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01292">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01292/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01292/status.svg
   :target: https://doi.org/10.21105/joss.01292
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01292/status.svg)](https://doi.org/10.21105/joss.01292)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-joss
